woocommerce/woocommerce-gateway-stripe
# Fix 'No such customer' error in new checkout experience

Fixes #3109 

Unable to place test orders after enabling the new checkout experience. The following error occurs in the checkout

![notice](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/b807d8ef-c785-4d88-bb3c-54ad00796da4)

This happens when the user tries to checkout with an email that had been previously used in live mode. The error does not occur in the legacy checkout experience.

## Changes proposed in this Pull Request:
With UPE enabled when we get this 'No such customer' error from Stripe, we delete the customer ID from the user meta and order meta. However, in the retry request, we still send the same customer ID. For this, all 5 retry requests fail and we get the error notice on the checkout page.

When legacy checkout is enabled (non-UPE), after deleting the customer ID we make the retry request with a recreated customer ID. So the payment gets successful on the 2nd retry request. 

In this PR, we are doing the same for UPE.

## Testing instructions

1. Go to your Stripe settings page and uncheck the `Enable the legacy checkout experience` checkbox (enable UPE).
2. Enable test mode from your Stripe settings page.
3. In your `wp-admin` create a new user (`donald.duck@test.com`).
4. Go to your Stripe dashboard's customer page (https://dashboard.stripe.com/customers) and create a new customer with the same email (`donald.duck@test.com`).
5. From your database, find the user ID for this newly created user. From Stripe dashboard find the customer ID from URL.
6. Now in the `wp_usermeta` table of your DB add a new row where the `user_id` is your new user's id, `meta_key` is `wp__stripe_customer_id` and `meta_value` is the customer id from Stripe dashboard.
7. As a shopper try to place an order using the above email id (`donald.duck@test.com`).
8. In `develop` branch you should get the mentioned error.

`There was an error processing the payment: No such customer: '[Stripe customer ID]'; a similar object exists in live mode, but a test mode key was used to make this request.`

9. In this branch, the checkout should be successful. In your Stripe dashboard, a new customer should be created in the test mode with the above email id.
